Media: Add support for selection with pan gesture

kean opened the pull request:
Part of https://github.com/wordpress-mobile/WordPress-iOS/issues/21457

- Add support for selection using pan gesture
- Add a light overlay over selected items

To test the ordered selection, go to `SiteMediaViewController.swift` and replace line 122:

```swift
collectionViewController.setEditing(isEditing, allowsMultipleSelection: true)
```

with this:

```swift
collectionViewController.setEditing(isEditing, allowsMultipleSelection: true, isSelectionOrdered: true)
```

I’ll make it easier to test the picker once this is merged: https://github.com/wordpress-mobile/WordPress-iOS/pull/21681

## To test:

It's a bit hard to describe how it should work and especially interact with previous selections, so I recorded two videos instead. This one demonstrates how it works with unordered selection, which is relatively easy:

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/664ce044-1119-4515-b6d0-d6f2b1bb566e

Another tricky part was making sure it works correctly with the existing pan gesture recognizer on a scroll view, so that you can both scroll _and_ select:

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/84d63607-cfca-4b27-be6c-3661c6b05087


And here is the ordered selection:

https://github.com/wordpress-mobile/WordPress-iOS/assets/1567433/bc98b5c9-aac9-4074-a928-c30b9362affd


## Regression Notes
1. Potential unintended areas of impact: n/a (no production code changes)
2. What I did to test those areas of impact (or what existing automated tests I relied on): manual
3. What automated tests I added (or what prevented me from doing so): n/a

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

kean committed:
Add support for selection with pan gesture

kean committed:
Add overlay for selected views

kean committed:
Update the pan gesture recognizer to work correctly with the existing collection view gestures

momo-ozawa reviewed (commented):
✨ Great improvements! A few things:

### Incorrect video duration

<img src="https://github.com/wordpress-mobile/WordPress-iOS/assets/6711616/30be30fd-be4f-4dc6-aa27-63886de15e3e" width=200>

I think this is a regression. The video duration is 00:00 for all videos.

### Video duration text displayed above selection badge

Sometimes the badge is displayed below the video duration text. I actually can't reproduce this issue but was able to take a screenshot (see above).

One solution is to make sure the badge is always displayed on top. Another solution would be to hide the video duration text when the cell is selected - this is what the iOS Photos app does.

### Deselecting via pan gesture
A nice-to-have: supporting deselecting via the pan action after you've made a selection. Currently you can deselect by panning over selected photos, if you haven't lifted your finger off the screen after selecting photos in the first place. But this is a nice-to-have and certainly not a requirement.

kean committed:
Show search when opening media screen

kean committed:
Add support for deselection with pan gesture

kean committed:
Hide badge when duration label is updated

kean commented:
- Video duration works the same way in the production version: API does not return the `length` field – investigating; not a merge blocker
- Duration about badge: great suggestion. I implemented it the same way as it's done in Phots, as you recommended
- Deselecting: implemented!

momo-ozawa reviewed (approved):
Nice! I've confirmed that deselecting via pan gesture and hiding the video duration during selection works as expected. Thanks for addressing those issues 🙇‍♀️

momo-ozawa commented on WordPress/Classes/ViewRelated/Media/SiteMedia/Views/SiteMediaCollectionCell.swift:
```diff
@@ -74,12 +74,8 @@ final class SiteMediaCollectionCell: UICollectionViewCell, Reusable {
-
-        viewModel.$durationText
-            .sink { [weak self] in self?.didUpdateDurationText($0) }
+        viewModel.$badge.combineLatest(viewModel.$durationText)
```
✨
